### PR TITLE
Allow lock errors on kubeconfig files that are never written

### LINKED ISSUE #32606: kubeconfig lock files change behavior of KUBECONFIG
Previously, if KUBECONFIG pointed to an empty directory that you don't have permission to write to, config would safely ignore that directory in its load path.  Now, the lock file raises a permission denied error which prevents normal operations from passing.

To recreate:

```
sudo mkdir /foo
KUBECONFIG=/foo/foo.config go test ./pkg/kubectl/cmd/config
```

This is a regression and causes old clients to break when they upgrade to 1.4 - if you had a KUBECONFIG before that pointed to an empty dir, we still need it.

### LINKED ISSUE #67676: kubeconfig config set-context fails with mix of writable & non-writable entries in KUBECONFIG
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

with `KUBECONFIG=/home/pdbogen/.kube/config:/etc/kube/config`, `kubeconfig config set current-context sandbox` returns `error: open /etc/kube/config.lock: permission denied`

**What you expected to happen**:

kubeconfig should lock only `/home/pdbogen/.kube/config` and set the indicated setting there

**How to reproduce it (as minimally and precisely as possible)**:

```sh
$ mkdir readonly writable
$ chmod -w readonly
$ KUBECONFIG=$PWD/writable/config:$PWD/readonly/config kubectl config set current-context example
error: open /home/pdbogen/readonly/config.lock: permission denied
```

**Anything else we need to know?**:

I want to be able to provide my users a system-managed kubernetes config that describes various clusters. I would like them to be able to specify things like current-context easily, via `kubectl config`.

Even when `~/.kube/config` exists and contains a `current-context` setting, `kubectl config set current-context` still tries to lock /etc/kube/config.

**Environment**:
- Kubernetes version (use `kubectl version`): `Client Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.5-16", GitCommit:"32ac1c9073b132b8ba18aa830f46b77dcceb0723", GitTreeState:"clean", BuildDate:"2018-06-27T17:46:30Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}` (we build our own, currently can't use 1.11)
- Cloud provider or hardware configuration: n/a; but, Amazon EKS
- OS (e.g. from /etc/os-release): Ubuntu Xenial 16.04.5
- Kernel (e.g. `uname -a`): 4.4.0-1063-aws
- Install tools: na
- Others: na

### PR DESCRIPTION
**Pull request by markus1189**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

When `KUBECONFIG` contains a list of files, where some of them are *not writable*, currently `kubectl` first tries to lock *all* of them, which leads to errors even if none of the files are consequently written.

**Which issue(s) this PR fixes**:

Fixes #67676
Fixes #32606

**Special notes for your reviewer**:
I'm not a GO programmer at all, so please take that into account during the review :)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go [hunk 1]
```diff
@@ -160,9 +160,12 @@ func ModifyConfig(configAccess ConfigAccess, newConfig clientcmdapi.Config, rela
...
 	sort.Strings(possibleSources)
+
+	var lockErrors map[string]error = make(map[string]error)
+
 	for _, filename := range possibleSources {
 		if err := lockFile(filename); err != nil {
-			return err
+			lockErrors[filename] = err
 		}
 		defer unlockFile(filename)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/config.go]: put this in an else block, we don't want to unlock files we did not actually lock

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go [hunk 2]
```diff
@@ -214,6 +217,10 @@ func ModifyConfig(configAccess ConfigAccess, newConfig clientcmdapi.Config, rela
 				}
 			}
 
+			if lockErrors[destinationFile] != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/config.go]: move this above the getConfigFromFile call... we want to make sure no one else is writing before we read in the file to get the current content we're going to rewrite

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go [hunk 3]
```diff
@@ -383,6 +390,11 @@ func writeCurrentContext(configAccess ConfigAccess, newCurrentContext string) er
 			return err
 		}
 		currConfig.CurrentContext = newCurrentContext
+
+		if lockErrors[file] != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/config.go]: move this above the getConfigFromFile call

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go [hunk 4]
```diff
@@ -398,6 +410,11 @@ func writeCurrentContext(configAccess ConfigAccess, newCurrentContext string) er
 		}
 		config.CurrentContext = newCurrentContext
 
+		if lockErrors[destinationFile] != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/config.go]: move this above the getConfigFromFile call

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go [hunk 5]
```diff
@@ -415,6 +432,11 @@ func writeCurrentContext(configAccess ConfigAccess, newCurrentContext string) er
 
 			if len(currConfig.CurrentContext) > 0 {
 				currConfig.CurrentContext = newCurrentContext
+
+				if lockErrors[file] != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/config.go]: move this above the getConfigFromFile call

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go [hunk 6]
```diff
@@ -456,6 +478,11 @@ func writePreferences(configAccess ConfigAccess, newPrefs clientcmdapi.Preferenc
 
 		if !reflect.DeepEqual(currConfig.Preferences, newPrefs) {
 			currConfig.Preferences = newPrefs
+
+			if lockErrors[file] != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/config.go]: move this above the getConfigFromFile call

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: and also add before the getConfigFromFile in the IsExplicitFile() block above

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go (hunk 1 above)
Comment (markus1189) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: Fixed in ecf3b83e88ef4c4791889f922514cf459d3b3058

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go (hunk 2 above)
Comment (markus1189) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: Sure you're totally right: Fixed in ecf3b83e88ef4c4791889f922514cf459d3b3058

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go (hunk 3 above)
Comment (markus1189) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: Fixed in ecf3b83e88ef4c4791889f922514cf459d3b3058

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go (hunk 4 above)
Comment (markus1189) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: Fixed in ecf3b83e88ef4c4791889f922514cf459d3b3058

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go (hunk 5 above)
Comment (markus1189) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: Fixed in ecf3b83e88ef4c4791889f922514cf459d3b3058

File: staging/src/k8s.io/client-go/tools/clientcmd/config.go (hunk 6 above)
Comment (markus1189) [reply staging/src/k8s.io/client-go/tools/clientcmd/config.go]: Done and done: ecf3b83e88ef4c4791889f922514cf459d3b3058

**Review by markus1189 — COMMENTED** [review]
I did sign the Linux CLA, however it still has the failing check...

**Comment by markus1189** [comment]
Also, I'm a little lost on how and where to best test this, so that we would notice a regression

**Comment by markus1189** [comment]
I would be willing to pick this up again if anybody would help me with the issues above

**Comment by alejandrox1** [comment]
> Also, I'm a little lost on how and where to best test this, so that we would notice a regression

@markus1189 from the looks of it, either compiling kubectl and trying to replicate the issue you're trying to fix or by using this patch to create a Kubernetes client and go through the motions of creating a KUBECONFIG containing a list of files, where some of them are not writable and checking if you get any errors should be a good way of testing this before CI, no?

**Review by alejandrox1 — COMMENTED** [review]
some nits

also, let's try this again
/check-cla

**Comment by markus1189** [comment]
I signed the CLA again :roll_eyes: Hopefully it works now.  As for the tests, it seems unlikely that this caused the performance tests to fail :thinking:  After clicking on `details` I did not find the exact log, help?

**Comment by markus1189** [comment]
> > Also, I'm a little lost on how and where to best test this, so that we would notice a regression
> 
> @markus1189 from the looks of it, either compiling kubectl and trying to replicate the issue you're trying to fix or by using this patch to create a Kubernetes client and go through the motions of creating a KUBECONFIG containing a list of files, where some of them are not writable and checking if you get any errors should be a good way of testing this before CI, no?

That's what I did, I compiled it and made sure that it works.  I thought about an automated test :)

**Comment by markus1189** [comment]
Sorry the rebase went wrong somehow, fixed now.  Still not sure about why the tests fail

**Comment by markus1189** [comment]
Closing because nobody seems to care and I myself don't care enough any more... :disappointed:

**Pull request closed without merging**